### PR TITLE
http: clear hop by hop TE header

### PR DESCRIPTION
**Pull request by alyssawilk**
Changes Envoy to clear TE header by default, as it is a hop by hop header and should not be forwarded on.

Risk Level: medium (behavioral change)
Testing: new e2e test
Docs Changes: n/a
Release Notes: inline.
[Optional Runtime guard:] yes.
Fixes https://github.com/envoyproxy/envoy/issues/30362

### CODE REVIEW

File: test/integration/header_integration_test.cc [hunk 1]
```diff
@@ -1375,8 +1377,9 @@ TEST_P(HeaderIntegrationTest, TestTeHeaderPassthrough) {
       });
 }
 
-// Validates TE header is stripped if it contains an unsupported value
+// Validates TE header was stripped if it contains an unsupported value
```
Comment (RyanTheOptimist) [inline test/integration/header_integration_test.cc:1380]: nit: Your call, but I'd be inclined to use the present tense here as it's describing what the test is *going* to do, not what it has done. But totally your call.

File: test/integration/header_integration_test.cc [hunk 2]
```diff
@@ -1341,8 +1341,10 @@ TEST_P(HeaderIntegrationTest, PathWithEscapedSlashesRedirected) {
                                    });
 }
 
-// Validates TE header is forwarded if it contains a supported value
+// Validates TE header was forwarded if it contains a supported value
+// It is now treated as the hop by hop header it is.
```
Comment (RyanTheOptimist) [inline test/integration/header_integration_test.cc:1345]: Should we remove this comment? I'm not sure that I understand why it's here?

File: changelogs/current.yaml [hunk 3]
```diff
@@ -2,6 +2,10 @@ date: Pending
 
 behavior_changes:
 # *Changes that are expected to cause an incompatibility if applicable; deployment changes are likely required*
+- area: http
+  change: |
+    Sanitize the hop by hop TE header from downstream request headers. This change can be temporarily reverted
+    by setting ``envoy.reloadable_features.sanitize_te`` to false.
```
Comment (RyanTheOptimist) [inline changelogs/current.yaml:8]: Does this flag *sanitize* the TE header, or does it *remove* it?

File: test/integration/protocol_integration_test.cc [hunk 4]
```diff
@@ -784,6 +784,29 @@ TEST_P(DownstreamProtocolIntegrationTest, MissingHeadersLocalReplyWithBodyBytesC
...
 
+TEST_P(DownstreamProtocolIntegrationTest, TeSanitization) {
+  autonomous_upstream_ = true;
+  config_helper_.addRuntimeOverride("envoy.reloadable_features.sanitize_te", "true");
+
+  if (downstreamProtocol() == Http::CodecType::HTTP1) {
+    default_request_headers_.setTE("gzip");
+  } else {
+    return;
+  }
```
Comment (RyanTheOptimist) [inline test/integration/protocol_integration_test.cc:795]: nit: Early return: 

```
if (downstreamProtocol() != Http::CodecType::HTTP1) {
  return;
}
default_request_headers_.setTE("gzip");
```

Optionally put this early return at the top of the test to make it clear that it's H1-specific.

**Review by RyanTheOptimist — COMMENTED** [review]
Looks great!

**Review by mattklein123 — DISMISSED** [review]

**Comment by quantumsheep** [comment]
Since Envoy 1.29 I get `Missing :te header` when proxying to a Python-gRPC API because of [this](https://github.com/grpc/grpc/blob/038215b504b9027ac85527f5fdcd85c76b7e3a1f/src/core/ext/filters/http/server/http_server_filter.cc#L102).

Removing the TE isn't the solution. We should replace it by `trailers`.

**Comment by alyssawilk** [comment]
So thankfully this is runtime guarded so you should be able to reinstate prior behavior with a config change.

I think if you want the envoy-upstream hop to support trailers you should just add request headers "te: trailers" to the outgoing request.

**Comment by ejona86** [comment]
> I think if you want the envoy-upstream hop to support trailers you should just add request headers "te: trailers" to the outgoing request.

If you are meaning unconditionally, that seems a bit simplistic. That ignores whether the originating client supports trailers. From RFC 7230 §4.1.2:

> The presence of the keyword "trailers" indicates that the client is
   willing to accept trailer fields in a chunked transfer coding, as
   defined in [Section 4.1.2](https://datatracker.ietf.org/doc/html/rfc7230#section-4.1.2), on behalf of itself and any downstream
   clients.

"any downstream clients" being the important part. So you'd want to make it conditional based on the originating request.

**Comment by RyanTheOptimist** [comment]
> > I think if you want the envoy-upstream hop to support trailers you should just add request headers "te: trailers" to the outgoing request.
> 
> If you are meaning unconditionally, that seems a bit simplistic. That ignores whether the originating client supports trailers. From RFC 7230 §4.1.2:
> 
> > The presence of the keyword "trailers" indicates that the client is
> > willing to accept trailer fields in a chunked transfer coding, as
> > defined in [Section 4.1.2](https://datatracker.ietf.org/doc/html/rfc7230#section-4.1.2), on behalf of itself and any downstream
> > clients.
> 
> "any downstream clients" being the important part. So you'd want to make it conditional based on the originating request.

Oh, interesting. Yes, I think that's a good point. Envoy itself is able to process trailers and pass them along, but it's only "safe" to do that in the case where Envoy knows that the downstream client is able to process trailers. The client, of course, indicates this by sending "trailers" in te. 

Does that sound right?

**Pull request merged**